mobile_verifier/src/speedtests.rs

```diff
@@ -96,580 +87,117 @@ impl SpeedtestDaemon {
-        }
-
-        self.file_sink.commit().await?;
-        transaction.commit().await?;
-
-        Ok(())
-    }
-}
-
-impl From<CellSpeedtest> for Speedtest {
-    fn from(cell_speedtest: CellSpeedtest) -> Self {
-        Self {
-            timestamp: cell_speedtest.timestamp,
-            upload_speed: cell_speedtest.upload_speed as i64,
-            download_speed: cell_speedtest.download_speed as i64,
-            latency: cell_speedtest.latency as i32,
-        }
-    }
-}
-
-impl PgHasArrayType for Speedtest {
-    fn array_type_info() -> PgTypeInfo {
-        PgTypeInfo::with_name("_speedtest")
-    }
-}
-
-#[derive(FromRow)]
-pub struct SpeedtestRollingAverage {
-    pub id: PublicKeyBinary,
-    pub speedtests: Vec<Speedtest>,
-    pub latest_timestamp: DateTime<Utc>,
-}
-
-impl SpeedtestRollingAverage {
-    pub fn new(id: PublicKeyBinary) -> Self {
-        Self {
-            id,
-            speedtests: Vec::new(),
-            latest_timestamp: DateTime::<Utc>::default(),
-        }
-    }
-
-    pub async fn validate_speedtests<'a>(
-        gateway_client: &'a GatewayClient,
-        speedtests: impl Stream<Item = CellSpeedtest> + 'a,
-        exec: &mut Transaction<'_, Postgres>,
-    ) -> Result<impl Stream<Item = Result<Self, ClientError>> + 'a, sqlx::Error> {
-        let tests_by_publickey = speedtests
-            .fold(
-                HashMap::<PublicKeyBinary, Vec<CellSpeedtest>>::new(),
-                |mut map, cell_speedtest| async move {
-                    map.entry(cell_speedtest.pubkey.clone())
-                        .or_default()
-                        .push(cell_speedtest);
-                    map
-                },
-            )
-            .await;
-
-        let mut speedtests = Vec::new();
-        for (pubkey, cell_speedtests) in tests_by_publickey.into_iter() {
-            let rolling_average: SpeedtestRollingAverage =
-                sqlx::query_as::<_, SpeedtestRollingAverage>(
-                    "SELECT * FROM speedtests WHERE id = $1",
-                )
-                .bind(&pubkey)
-                .fetch_optional(&mut *exec)
-                .await?
-                .unwrap_or_else(|| SpeedtestRollingAverage::new(pubkey.clone()));
-            speedtests.push((rolling_average, cell_speedtests));
-        }
-
-        Ok(futures::stream::iter(speedtests.into_iter())
-            .then(move |(rolling_average, cell_speedtests)| {
-                async move {
-                    // If we get back some gateway info for the given address, it's a valid address
-                    if gateway_client
-                        .resolve_gateway_info(&rolling_average.id)
-                        .await?
-                        .is_none()
-                    {
-                        return Ok(None);
-                    }
-                    Ok(Some((rolling_average, cell_speedtests)))
+            .map(anyhow::Ok)
+            .try_fold(transaction, |mut transaction, report| async {
```

Comment:
I think that using a try fold, and separating out the functions is less clear and incongruous with how heartbeats (and eventually coverage object) daemons are structured, this function should match how the heartbeats function operates. That is, have a `validate_speedtests` function on `Speedtest` that you can call `write` and `save` on in a while loop.  

Basically, I think that the structure of the function before doesn't need to be changed, because it matches the way heartbeats work and how the coverage object work. Changing this function to look differently makes the function less clear